Add lidar modularity for ROS2 Gem

alek-kam-robotec-ai opened the pull request:
This pull request introduces lidar implementation modularity by allowing for multiple LiDAR implementations to be added and used by the ROS2LidarSensor. This can be done by implementing the Lidar System and Raycaster interfaces either in the form of a new Gem or within the ROS Gem.  

Only manual tests of the new Lidar functionalities have been done.

michalpelka commented on Gems/ROS2/Code/Source/Lidar/LidarRegistrarSystemComponent.h:
```diff
@@ -0,0 +1,49 @@
+        void Activate() override;
+        void Deactivate() override;
+
+        ////////////////////////////////////////////////////////////////////////
```
```suggestion
        ////////////////////////////////////////////////////////////////////////
        // LidarRegistrarRequestBus
```

alek-kam-robotec-ai replied:
Thanks! Fixed.

michalpelka commented on Gems/ROS2/Code/Source/Lidar/LidarSystem.cpp:
```diff
@@ -0,0 +1,58 @@
+
+    void LidarSystem::Activate()
+    {
+        const char* name = "PhysX";
```
Const name should be capitalized : https://github.com/o3de/sig-core/blob/main/governance/Coding-Standards-and-Style-Guide.md#constant-names

alek-kam-robotec-ai replied:
Thanks! Fixed.

michalpelka commented on Gems/ROS2/Code/Source/Lidar/LidarTemplateUtils.cpp:
```diff
@@ -59,23 +165,38 @@ namespace ROS2
-                const float pitch = minVertAngle + layer * verticalStep + rootRotation.GetY();
-                const float yaw = minHorAngle + incr * horizontalStep + rootRotation.GetZ();
+                const float pitch = minVertAngle + layer * verticalStep;
+                const float yaw = minHorAngle + incr * horizontalStep + (AZ::Constants::Pi / 2.0f);
```
```suggestion
                const float yaw = minHorAngle + incr * horizontalStep + (AZ::Constants:HalfPi);
```

alek-kam-robotec-ai replied:
Thanks! Fixed.

michalpelka commented on Gems/ROS2/Code/Source/Lidar/ROS2LidarSensorComponent.cpp:
```diff
@@ -133,13 +230,25 @@ namespace ROS2
+
+        m_lastRotations = LidarTemplateUtils::PopulateRayRotations(m_lidarParameters);
+
+        if (m_lidarSystem == AZStd::string(""))
```
Consider
```suggestion 
        if (m_lidarSystem.empty())
```

alek-kam-robotec-ai replied:
Thanks! Fixed.

michalpelka commented on Gems/ROS2/Code/ros2_files.cmake:
```diff
@@ -28,6 +28,10 @@ set(FILES
         Source/Imu/ROS2ImuSensorComponent.h
         Source/Lidar/LidarRaycaster.cpp
```
Please keep alphabetical order.
```suggestion
        Source/Lidar/LidarRaycaster.cpp
        Source/Lidar/LidarRaycaster.h
        Source/Lidar/LidarRegistrarSystemComponent.cpp
        Source/Lidar/LidarRegistrarSystemComponent.h
        Source/Lidar/LidarSystem.cpp
        Source/Lidar/LidarSystem.h
````

michalpelka reviewed (commented):
Left some nitpick comments from my side.

hultonha commented on Gems/ROS2/Code/Include/ROS2/Lidar/LidarRaycasterBus.h:
```diff
@@ -0,0 +1,104 @@
+
+namespace ROS2
+{
+    class LidarRaycasterRequests
```
suggestion: It'd be helpful to add API comments for this public type to help future maintainers (and the docs generation I'm promised is still coming 😅).

alek-kam-robotec-ai replied:
Thanks! Fixed.

hultonha commented on Gems/ROS2/Code/Include/ROS2/Lidar/LidarRaycasterBus.h:
```diff
@@ -0,0 +1,104 @@
+    {
+    public:
+        AZ_RTTI(LidarRaycasterRequests, "{253a02c8-b6cb-493c-b16f-012ccf9db226}");
+        virtual ~LidarRaycasterRequests() = default;
```
suggestion: As we never intend to delete an instance through the `LidarRaycasterRequests` it's actually preferred to make this protected and non-virtual (most request bus interfaces try and take this approach now)

alek-kam-robotec-ai replied:
Thanks! Fixed.

hultonha commented on Gems/ROS2/Code/Include/ROS2/Lidar/LidarRaycasterBus.h:
```diff
@@ -0,0 +1,104 @@
+
+        //! Configures ray Gaussian Noise parameters.
+        //! Each call overrides the previous configuration.
+        //! This type of noise is especially useful when trying to simulate real-life lidars, since it noise mimics
```
uber nit: it noise -> its noise

alek-kam-robotec-ai replied:
Thanks! Fixed.

hultonha commented on Gems/ROS2/Code/Include/ROS2/Lidar/LidarRaycasterBus.h:
```diff
@@ -0,0 +1,104 @@
+        //! The returned vector size can be anything between zero and size of directions. No hits further than distance will be reported.
+        virtual AZStd::vector<AZ::Vector3> PerformRaycast(const AZ::Transform& lidarTransform) = 0;
+
+        //! Configures ray Gaussian Noise parameters.
```
These comments are excellent, super detailed and helpful for the reader, great to see! 👍

alek-kam-robotec-ai replied:
Thanks a lot!

hultonha commented on Gems/ROS2/Code/Include/ROS2/Lidar/LidarRaycasterBus.h:
```diff
@@ -0,0 +1,104 @@
+        //! Configures Layer ignoring parameters
+        //! @param ignoreLayer Should a specified collision layer be ignored?
+        //! @param layerIndex Index of collision layer to be ignored.
+        virtual void ConfigureLayerIgnoring(bool ignoreLayer, unsigned int layerIndex)
```
nit: Prefer `AZ::u32` here rather than `unsigned int`

alek-kam-robotec-ai replied:
Thanks! Fixed.

hultonha commented on Gems/ROS2/Code/Include/ROS2/Lidar/LidarRaycasterBus.h:
```diff
@@ -0,0 +1,104 @@
+        }
+
+        //! Excludes entities with given EntityIds from raycasting.
+        //! @param excludedEntities list of entities marked for exclusion.
```
uber nit: list -> List

alek-kam-robotec-ai replied:
Thanks! Fixed.

hultonha commented on Gems/ROS2/Code/Include/ROS2/Lidar/LidarRaycasterBus.h:
```diff
@@ -0,0 +1,104 @@
+        }
+
+    protected:
+        static void ValidateRayRange(float range)
+        {
+            AZ_Assert(range > 0.0f, "Provided ray range was of incorrect value: Ray range value must be greater than zero.")
+        }
+
+        static void ValidateRayOrientations(const AZStd::vector<AZ::Vector3>& orientations)
+        {
+            AZ_Assert(!orientations.empty(), "Provided ray orientations were of incorrect value: Ray orientations must not be empty.")
+        }
```
question: Is having these static functions on the interface a good idea? Should these just be implemented in the derived type? Not sure 100% how this will play with the EBus mechanics

alek-kam-robotec-ai replied:
These restrictions are common to all lidar implementations which if implemented in derived types would add additional boilerplate to each Lidar implementation. If however they could cause issues with the EBus mechanics I'm open to removing them from the interface.

hultonha commented on Gems/ROS2/Code/Include/ROS2/Lidar/LidarRegistrarBus.h:
```diff
@@ -0,0 +1,72 @@
+    //! Structure used to describe LidarSystem's feature support.
+    struct LidarSystemFeatures
+    {
+    public:
```
nit: `public` can be omitted as this is a struct and public be default

alek-kam-robotec-ai replied:
Thanks! Fixed.

hultonha commented on Gems/ROS2/Code/Include/ROS2/Lidar/LidarRegistrarBus.h:
```diff
@@ -0,0 +1,72 @@
+    //! Structure used to hold LidarSystem's metadata.
+    struct LidarSystemMetaData
+    {
+    public:
```
update: Please remove `public`

alek-kam-robotec-ai replied:
Thanks! Fixed.

hultonha commented on Gems/ROS2/Code/Include/ROS2/Lidar/LidarRegistrarBus.h:
```diff
@@ -0,0 +1,72 @@
+    {
+    public:
+        AZ_RTTI(LidarRegistrarRequests, "{22030dc7-a1db-43bd-b748-0fb9ec43ce2e}");
+        virtual ~LidarRegistrarRequests() = default;
```
nit: Same comment about destructor as before

alek-kam-robotec-ai replied:
Thanks! Fixed.

hultonha commented on Gems/ROS2/Code/Include/ROS2/Lidar/LidarRegistrarBus.h:
```diff
@@ -0,0 +1,72 @@
+
+        //! Returns a list of all registered lidar systems.
+        //! @return a vector of registered lidar systems' names.
+        virtual const AZStd::vector<AZStd::string> GetRegisteredLidarSystems() = 0;
```
nit: As this vector is being returned by value please remove the `const` (or make this a reference if that's what's intended)

alek-kam-robotec-ai replied:
Thanks! Fixed.

hultonha commented on Gems/ROS2/Code/Include/ROS2/Lidar/LidarSystemBus.h:
```diff
@@ -0,0 +1,39 @@
+    {
+    public:
+        AZ_RTTI(LidarSystemRequests, "{007871d1-2783-4382-977b-558f436c54a5}");
+        virtual ~LidarSystemRequests() = default;
```
nit: Destructor comment from before

alek-kam-robotec-ai replied:
Thanks! Fixed.

hultonha commented on Gems/ROS2/Code/Include/ROS2/Lidar/LidarSystemBus.h:
```diff
@@ -0,0 +1,39 @@
+        //! Creates a new Lidar.
+        //! @param lidarEntityId EntityId holding the ROS2LidarSensorComponent.
+        //! @return A unique Id of the newly created Lidar.
+        virtual AZ::Uuid CreateLidar(const AZ::EntityId& lidarEntityId) = 0;
```
suggestion: Creating a typesafe version of `AZ::Uuid` might be a useful addition (could be outside the scope of this PR).

moraaar replied:
I like Tom suggestion to do a typesafe version of `AZ::Uuid` for Lidar. Or as minimum an alias would be nice to make it easier to change it in the future.

````
using LidarId = AZ::Uuid;
````

That way `LidarId` could be used everywhere.

alek-kam-robotec-ai replied:
Thanks! I've created the `LidarId` alias as you had suggested (inside the LidarRaycasterRequests class).

hultonha replied:
This is definitely better but it's not type safe unfortunately (take a look at how we create type safe ids in the White Box Gem if you're interested (see WhiteBoxToolApi.h and look for all the handle types - e.g. VertexHandle, EdgeHandle etc... It makes it impossible to accidentally mix them up which is super useful (runtime bugs become compile time bugs 🥳). It uses this technique for reference - https://blog.demofox.org/2015/02/05/getting-strongly-typed-typedefs-using-phantom-types/.

alek-kam-robotec-ai replied:
Thanks for this suggestion. I've updated this section of code. The references you provided were very helpful!

hultonha replied:
No problem! Glad it was useful 🙂

hultonha commented on Gems/ROS2/Code/Include/ROS2/Lidar/LidarSystemBus.h:
```diff
@@ -0,0 +1,39 @@
+        //! Creates a new Lidar.
+        //! @param lidarEntityId EntityId holding the ROS2LidarSensorComponent.
+        //! @return A unique Id of the newly created Lidar.
+        virtual AZ::Uuid CreateLidar(const AZ::EntityId& lidarEntityId) = 0;
```
nit: As EntityIds are just a wrapper around a `u64` it's better to just pass by value so this can `EntityId lidarEntityId`

alek-kam-robotec-ai replied:
Thanks! Fixed.

hultonha commented on Gems/ROS2/Code/Source/Lidar/LidarRaycaster.h:
```diff
@@ -15,10 +16,14 @@
+    class LidarRaycaster : protected LidarRaycasterRequestBus::Handler
     {
     public:
+        explicit LidarRaycaster(const LidarRaycasterRequestBus::BusIdType& busId);
+        LidarRaycaster(LidarRaycaster&& lidarSystem) noexcept;
+        LidarRaycaster(const LidarRaycaster& lidarSystem) = delete;
+        ~LidarRaycaster() override;
```
suggestion: Rule of 5 is good to follow here so we likely want to provide move assignment and copy assignment operators too otherwise this type could break in certain situations

alek-kam-robotec-ai replied:
Thanks! Fixed.

hultonha commented on Gems/ROS2/Code/Source/Lidar/LidarRaycaster.h:
```diff
@@ -27,29 +32,25 @@ namespace ROS2
-        //! If true the raycaster will also include points at maximum range when nothing was hit
-        void SetAddPointsMaxRange(bool addPointsMaxRange);
+    protected:
+        ////////////////////////////////////////////////////////////////////////
```
nit: We tend to recommend avoiding these blocks these days as they mostly just add visual noise/clutter

alek-kam-robotec-ai replied:
Thanks! Fixed.

hultonha commented on Gems/ROS2/Code/Source/Lidar/LidarRaycaster.h:
```diff
@@ -27,29 +32,25 @@ namespace ROS2
-        void SetAddPointsMaxRange(bool addPointsMaxRange);
+    protected:
+        ////////////////////////////////////////////////////////////////////////
+        // LidarRaycasterRequestBus::Handler interface implementation
```
nit: Prefer `LidarRaycasterRequestBus overrides ...`

alek-kam-robotec-ai replied:
Thanks! Fixed.

hultonha commented on Gems/ROS2/Code/Source/Lidar/LidarRaycaster.h:
```diff
@@ -27,29 +32,25 @@ namespace ROS2
+        AZStd::vector<AZ::Vector3> m_rayRotations{ { AZ::Vector3::CreateZero() } };
+
+        bool m_ignoreLayer{ false };
+        unsigned int m_ignoredLayerIndex{ 0 };
```
nit: Prefer `AZ::u32`

alek-kam-robotec-ai replied:
Thanks! Fixed.

hultonha commented on Gems/ROS2/Code/Source/Lidar/LidarRegistrarSystemComponent.cpp:
```diff
@@ -0,0 +1,108 @@
+{
+    LidarRegistrarSystemComponent::LidarRegistrarSystemComponent()
+    {
+        if (LidarRegistrarInterface::Get() == nullptr)
```
uber nit: `if (!LidarRegistrarInterface::Get())` style is generally preferred

alek-kam-robotec-ai replied:
Thanks! Fixed.

hultonha commented on Gems/ROS2/Code/Source/Lidar/LidarSystem.cpp:
```diff
@@ -0,0 +1,58 @@
+        const char* name = "PhysX";
+        const char* description = "Collider-based lidar implementation that uses the PhysX engine's raycasting.";
+        const LidarSystemFeatures supportedFeatures = {
+            /* .m_noise =                   */ false,
+            /* .m_collisionLayers =         */ true,
+            /* .m_entityExclusion =         */ false,
+            /* .m_MaxRangeHitPointConfig =  */ true,
```
Thanks for adding these comments, I wish we had C designated initializer support in C++ (apparently 20 has it so maybe one day) - the other option that can sometimes be nicer and easier to read is to use enum bitwise flags.

alek-kam-robotec-ai replied:
Thanks! The enum bitwise flags are a great idea. Done.

hultonha commented on Gems/ROS2/Code/Source/Lidar/LidarSystem.cpp:
```diff
@@ -0,0 +1,58 @@
+        m_lidars.back().SetRaycasterScene(GetPhysicsSceneFromEntityId(lidarEntityId));
+        return lidarUuid;
+    }
+} // namespace ROS2
```
nit: EOL missing

alek-kam-robotec-ai replied:
Thanks! Fixed.

hultonha commented on Gems/ROS2/Code/Source/Lidar/LidarSystem.h:
```diff
@@ -0,0 +1,34 @@
+
+    private:
+        ////////////////////////////////////////////////////////////////////////
+        // LidarSystemRequestBus::Handler interface implementation
```
nit: Same API interface comments as before

alek-kam-robotec-ai replied:
Thanks! Fixed.

hultonha commented on Gems/ROS2/Code/Source/Lidar/LidarTemplateUtils.cpp:
```diff
@@ -18,21 +18,128 @@ namespace ROS2
+                },
+            },
+            {
+                LidarTemplate::LidarModel::Velodyne_HDL_32E,
+                {
+                    /*.m_model = */ LidarTemplate::LidarModel::Velodyne_HDL_32E,
+                    /*.m_name = */ "Velodyne HDL-32E",
+                    /*.m_minHAngle = */ -180.0f,
+                    /*.m_maxHAngle = */ 180.0f,
+                    /*.m_minVAngle = */ 10.67f,
+                    /*.m_maxVAngle = */ -30.67f,
+                    /*.m_layers = */ 32,
+                    /*.m_numberOfIncrements = */ 1800, // For 0.2 angular resolution
+                    /*.m_maxRange = */ 100.0f,
+                    /*.m_noiseParameters = */
+                    {
+                        /*.m_angularNoiseStdDev = */ 0.0f,
+                        /*.m_distanceNoiseStdDevBase = */ 0.02f,
+                        /*.m_distanceNoiseStdDevRisePerMeter = */ 0.001f,
+                    },
```
+1 to this style I like it! As a stand-in for designated initializers it's not bad! 🙂

hultonha reviewed (approved):
Not a SME on this area I'm afraid but had a quick look through and it overall looks good to me! Nice work @alek-kam-robotec-ai 👍

alek-kam-robotec-ai commented on Gems/ROS2/Code/Source/Lidar/ROS2LidarSensorComponent.cpp:
```diff
@@ -133,13 +230,25 @@ namespace ROS2
+            AZStd::vector<AZStd::string> lidarSystemList = GetLidarSystemList();
+            AZ_Assert(!lidarSystemList.empty(), "Unable to choose a lidarSystem for the lidar sensor.");
+
+            m_lidarSystem = lidarSystemList.at(0);
```
Thanks! Fixed.

alek-kam-robotec-ai commented on Gems/ROS2/Code/Source/Lidar/ROS2LidarSensorComponent.cpp:
```diff
@@ -78,6 +132,64 @@ namespace ROS2
+
+    void ROS2LidarSensorComponent::ConnectToLidarRaycaster()
+    {
+        auto raycasterId = m_implementationToRaycasterMap.find(m_lidarSystem);
+        if (raycasterId != m_implementationToRaycasterMap.end())
```
Thanks! Fixed.

alek-kam-robotec-ai commented on Gems/ROS2/Code/Source/Lidar/LidarSystem.h:
```diff
@@ -0,0 +1,34 @@
+    {
+    public:
+        LidarSystem() = default;
+        LidarSystem(LidarSystem&& lidarSystem) = delete;
+        LidarSystem(const LidarSystem& lidarSystem) = delete;
```
Thanks! Fixed.

alek-kam-robotec-ai commented on Gems/ROS2/Code/Include/ROS2/Lidar/LidarRaycasterBus.h:
```diff
@@ -0,0 +1,104 @@
+        //! @param distanceNoiseStdDevBase Base value for Distance noise standard deviation.
+        //! @param distanceNoiseStdDevRisePerMeter Value by which the distance noise standard deviation increases per meter length from
+        //! the lidar.
+        // TODO - different starting points for rays, distance from reference point, noise models, rotating mirror sim, other
```
Thanks! Fixed.

hultonha commented:
Give me a poke @alek-kam-robotec-ai when you'd like me to kick off a build 👍

alek-kam-robotec-ai commented on Gems/ROS2/Code/Include/ROS2/Lidar/LidarSystemBus.h:
```diff
@@ -0,0 +1,39 @@
+    };
+
+    using LidarSystemRequestBus = AZ::EBus<LidarSystemRequests, LidarSystemBusTraits>;
+} // namespace ROS2
```
Thanks! Fixed.

alek-kam-robotec-ai commented on Gems/ROS2/Code/Source/Lidar/LidarRaycaster.cpp:
```diff
@@ -16,40 +17,83 @@
+        ROS2::LidarRaycasterRequestBus::Handler::BusConnect(busId);
+    }
+
+    LidarRaycaster::LidarRaycaster(LidarRaycaster&& lidarRaycaster) noexcept
```
Thanks! Fixed.

alek-kam-robotec-ai commented on Gems/ROS2/Code/Source/Lidar/LidarRaycaster.cpp:
```diff
@@ -16,40 +17,83 @@
+        m_range = range;
+    }
+
+    // A simplified, non-optimized first version. TODO - generalize results (fields)
```
Thanks! Fixed.

alek-kam-robotec-ai commented on Gems/ROS2/Code/Source/Lidar/LidarRaycaster.cpp:
```diff
@@ -16,40 +17,83 @@
-            request->m_filterCallback =
-                [ignoreLayer, ignoredLayerIndex](const AzPhysics::SimulatedBody* simBody, const Physics::Shape* shape)
+            request->m_filterCallback = [this](const AzPhysics::SimulatedBody* simBody, const Physics::Shape* shape)
             {
```
No, it is not. Thanks! Fixed it.

moraaar commented on Gems/ROS2/Code/Source/Lidar/LidarRaycaster.cpp:
```diff
@@ -16,40 +17,83 @@
             return results;
         }
 
+        AZStd::vector<AZ::Vector3> rayDirections =
```
`rayDirections` can be const.

alek-kam-robotec-ai replied:
Thanks! Done.

moraaar commented on Gems/ROS2/Code/Source/Lidar/LidarRaycaster.cpp:
```diff
@@ -16,40 +17,83 @@
+        AZStd::vector<AZ::Vector3> rayDirections =
+            LidarTemplateUtils::RotationsToDirections(m_rayRotations, lidarTransform.GetEulerRadians());
+
+        AZ::Vector3 lidarPosition = lidarTransform.GetTranslation();
```
`lidarPosition` can be const.

moraaar commented on Gems/ROS2/Code/Include/ROS2/ROS2Bus.h:
```diff
@@ -9,6 +9,7 @@
 
 #include <AzCore/EBus/EBus.h>
 #include <AzCore/Interface/Interface.h>
+#include <AzCore/std/string/string.h>
```
This file doesn't seem to reference `AZStd::string`, is it actually needed?

alek-kam-robotec-ai replied:
No, it's not. Thanks!

moraaar commented on Gems/ROS2/Code/Include/ROS2/Lidar/LidarRaycasterBus.h:
```diff
@@ -0,0 +1,104 @@
+        //! the lidar.
+        // TODO - different starting points for rays, distance from reference point, noise models, rotating mirror sim, other
+        // TODO - customized settings. Encapsulate in lidar definition and pass in constructor, update transform.
+        virtual void ConfigureNoiseParameters(
+            float angularNoiseStdDev, float distanceNoiseStdDevBase, float distanceNoiseStdDevRisePerMeter)
+        {
+            AZ_Assert(false, "This Lidar Implementation does not support noise!");
+        }
```
Please add `[[maybe_unused]]` to all the parameters to avoid compilation errors.

Same in all the other functions in this file that don't use their parameters.

alek-kam-robotec-ai replied:
Thanks! Fixed.

moraaar commented on Gems/ROS2/Code/Source/Lidar/LidarSystem.cpp:
```diff
@@ -0,0 +1,58 @@
+    AzPhysics::SceneHandle GetPhysicsSceneFromEntityId(const AZ::EntityId& entityId)
+    {
+        auto* physicsSystem = AZ::Interface<AzPhysics::SystemInterface>::Get();
+        auto foundBody = physicsSystem->FindAttachedBodyHandleFromEntityId(entityId);
```
When is the simulated body this is querying created?

It's important to think when this is asked and at what point they are created, otherwise this will always return invalid body.

From what I can see in the PR the chain of calls that arrive to here is as follows:

`ROS2LidarSensorComponent::Activate()` > `ConnectToLidarRaycaster()` > `CreateLidar(GetEntityId())` > `GetPhysicsSceneFromEntityId(entityId)` > FindAttachedBodyHandleFromEntityId(entityId)

So the body is expected to be already created when `ROS2LidarSensorComponent` is activated. 

When and where is the body created? Does it happen before  `ROS2LidarSensorComponent` is activated?

alek-kam-robotec-ai replied:
Yes, I see where problems could arise now. I've moved this function into the LidarRaycaster.cpp file and made it so the handle is set upon first raycast (which guarantees that the entity with said ROS2LidarSensor component is activated). Thanks for spotting that!

moraaar reviewed (commented):
Thanks for this contribution. I left some feedback.

It'd be great if the PR's description could contain more information about what the code is doing and an overall description of what's trying to do.

Also a section in the description indicating what testing (if any) has been done to check this change would be great.

Lastly, are there any unit tests for Lidar in ROS2 gem? If there are it'd be great to update them to cover this change. If no, is there any plans to add them later?

Thanks

alek-kam-robotec-ai commented on Gems/ROS2/Code/Source/Lidar/LidarTemplateUtils.cpp:
```diff
@@ -59,23 +165,38 @@ namespace ROS2
+        directions.reserve(rotations.size());
+        for (auto angle : rotations)
+        {
+            const float pitchTransformed = angle.GetY() + rootRotation.GetY();
+            const float yawTransformed = angle.GetZ() + rootRotation.GetZ();
+
+            const float x = AZ::Cos(yawTransformed) * AZ::Cos(pitchTransformed);
+            const float y = AZ::Sin(yawTransformed) * AZ::Cos(pitchTransformed);
+            const float z = AZ::Sin(pitchTransformed);
+
+            directions.emplace_back(x, y, z);
```
Yes, it is possible. Thanks!

hultonha replied:
Awesome stuff, thanks!

alek-kam-robotec-ai commented on Gems/ROS2/Code/Source/ROS2ModuleInterface.h:
```diff
@@ -7,6 +7,7 @@
  */
 #pragma once
 
+#include "Lidar/LidarRegistrarSystemComponent.h"
```
Thanks! Fixed.

alek-kam-robotec-ai commented on Gems/ROS2/Code/Source/Lidar/ROS2LidarSensorComponent.h:
```diff
@@ -40,15 +43,27 @@ namespace ROS2
+        bool IsIgnoredLayerConfigurationVisible() const;
+        bool IsEntityExclusionVisible() const;
+        bool IsMaxPointsConfigurationVisible() const;
+        AZStd::vector<AZStd::string> GetLidarSystemList();
```
Thanks! That's a good suggestion. Done.

alek-kam-robotec-ai commented on Gems/ROS2/Code/Source/Lidar/ROS2LidarSensorComponent.cpp:
```diff
@@ -16,6 +15,7 @@
 #include <Lidar/LidarTemplateUtils.h>
 #include <Lidar/ROS2LidarSensorComponent.h>
 #include <ROS2/Frame/ROS2FrameComponent.h>
+#include <ROS2/Lidar/LidarRegistrarBus.h>
```
Thanks! Fixed.

alek-kam-robotec-ai commented on Gems/ROS2/Code/Include/ROS2/Lidar/LidarRegistrarBus.h:
```diff
@@ -0,0 +1,72 @@
+    struct LidarSystemFeatures
+    {
+    public:
+        bool m_noise{ false };
+        bool m_collisionLayers{ false };
+        bool m_entityExclusion{ false };
+        bool m_maxRangePoints{ false };
```
Thanks, I agree. As Tom suggested this struct will be changed to a bit flag enum which these names should fit.

alek-kam-robotec-ai commented on Gems/ROS2/Code/Source/Lidar/ROS2LidarSensorComponent.cpp:
```diff
@@ -133,13 +230,25 @@ namespace ROS2
+
+        if (m_lidarSystem == AZStd::string(""))
+        {
+            AZStd::vector<AZStd::string> lidarSystemList = GetLidarSystemList();
```
Thanks! Fixed.

alek-kam-robotec-ai commented on Gems/ROS2/Code/Include/ROS2/Lidar/LidarRaycasterBus.h:
```diff
@@ -0,0 +1,104 @@
+        }
+
+    protected:
+        static void ValidateRayRange(float range)
```
These errors should never happen due to user actions. Yes. Fixed.

alek-kam-robotec-ai commented on Gems/ROS2/Code/Source/Lidar/LidarRaycaster.cpp:
```diff
@@ -63,26 +107,30 @@ namespace ROS2
+            else if (m_addMaxRangePoints)
             {
-                results.push_back(globalToLidarTM.TransformPoint(start + directions[i] * distance));
+                AZ::Vector3 maxPoint = lidarTransform.TransformPoint(rayDirections[i] * m_range);
```
Thanks! Fixed.

alek-kam-robotec-ai commented on Gems/ROS2/Code/Source/Lidar/LidarRaycaster.cpp:
```diff
@@ -16,40 +17,83 @@
+
+    LidarRaycaster::~LidarRaycaster()
+    {
+        if (!m_uuid.IsNull())
+        {
+            ROS2::LidarRaycasterRequestBus::Handler::BusDisconnect();
+        }
```
Thanks! Fixed that as well.

alek-kam-robotec-ai commented on Gems/ROS2/Code/Source/Lidar/LidarRegistrarEditorSystemComponent.cpp:
```diff
@@ -0,0 +1,53 @@
+
+    void LidarRegistrarEditorSystemComponent::GetProvidedServices(AZ::ComponentDescriptor::DependencyArrayType& provided)
+    {
+        BaseSystemComponent::GetProvidedServices(provided);
```
This piece of code was generated using a SystemComponent template and since the BaseSystemComponent is an alias for LidarRegistrarSystemComponent I saw no purpose in removing it (I'm not exactly sure why the SystemComponent template uses it in the first place). Throughout the engine many gems seem to either remove or leave it.

alek-kam-robotec-ai commented on Gems/ROS2/Code/Source/Lidar/LidarRegistrarSystemComponent.cpp:
```diff
@@ -0,0 +1,108 @@
+        return lidarSystemList;
+    }
+
+    const LidarSystemMetaData& LidarRegistrarSystemComponent::GetLidarSystemMetaData(const AZStd::string& name)
+    {
+        auto lidarSystem = m_registeredLidarSystems.find(AZ_CRC(name));
+        AZ_Assert(lidarSystem != m_registeredLidarSystems.end(), "No registered lidar system matches the provided name.");
+        return lidarSystem->second;
+    }
```
This is a good suggestion. Done.

alek-kam-robotec-ai commented on Gems/ROS2/Code/Source/Lidar/LidarRegistrarSystemComponent.cpp:
```diff
@@ -0,0 +1,108 @@
+    {
+        AZStd::vector<AZStd::string> lidarSystemList;
+        lidarSystemList.reserve(m_registeredLidarSystems.size());
+        for (auto lidarSystem : m_registeredLidarSystems)
```
Thanks! Fixed.

alek-kam-robotec-ai commented on Gems/ROS2/Code/Source/Lidar/LidarRegistrarSystemComponent.h:
```diff
@@ -0,0 +1,49 @@
+
+        ////////////////////////////////////////////////////////////////////////
+        void RegisterLidarSystem(const char* name, const char* description, const LidarSystemFeatures& features) override;
+        const AZStd::vector<AZStd::string> GetRegisteredLidarSystems() override;
+        const LidarSystemMetaData& GetLidarSystemMetaData(const AZStd::string& name) override;
```
Yes. Thanks!

alek-kam-robotec-ai commented on Gems/ROS2/Code/Source/Lidar/LidarRegistrarSystemComponent.cpp:
```diff
@@ -0,0 +1,108 @@
+    {
+    }
+
+    void LidarRegistrarSystemComponent::Activate()
+    {
+        m_physxLidarSystem.Activate();
+    }
+
+    void LidarRegistrarSystemComponent::Deactivate()
+    {
+    }
```
Thanks! Fixed.

moraaar reviewed (approved):
Great job @alek-kam-robotec-ai! A couple of minor comments, but nothing blocking. Approved.

alek-kam-robotec-ai commented on Gems/ROS2/Code/Include/ROS2/Lidar/LidarRegistrarBus.h:
```diff
@@ -0,0 +1,76 @@
+        CollisionLayers     = 0b00000010,
+        EntityExclusion     = 0b00000100,
+        MaxRangePoints      = 0b00001000,
+        All                 = 0b00001111
```
Thanks! I think that uint16_t should be enough.

hultonha commented:
Awesome stuff

> > Give me a poke @alek-kam-robotec-ai when you'd like me to kick off a build +1
> 
> @hultonha I've applied requested changes. Build should be ready to go. Thanks!

Awesome stuff, just kicked off a build @alek-kam-robotec-ai 👍 https://jenkins.build.o3de.org/view/Projects/job/o3de-extras/view/change-requests/job/PR-101/1/

alek-kam-robotec-ai commented:
> Awesome stuff
> 
> > > Give me a poke @alek-kam-robotec-ai when you'd like me to kick off a build +1
> > 
> > 
> > @hultonha I've applied requested changes. Build should be ready to go. Thanks!
> 
> Awesome stuff, just kicked off a build @alek-kam-robotec-ai +1 https://jenkins.build.o3de.org/view/Projects/job/o3de-extras/view/change-requests/job/PR-101/1/

Although the build has failed it is not clear to me why. The errors inside the pipeline.log file do not seem to be connected in any way to the gem's code. Similar build results can be seen in the [Introduced XACRO parsing to ROS2 Gem](https://github.com/o3de/o3de-extras/commit/7ca4fb7f58b25786a71fba17c9ae731c006c8f47) commit (the worklow run 2 hours before this one).

moraaar commented:
@byrcolin the AR for this [o3de-extra PR](https://jenkins.build.o3de.org/blue/organizations/jenkins/o3de-extras/detail/PR-101/1/pipeline/950) is failing to setup, is this something you can help with?

hultonha commented:
> > Awesome stuff
> > > > Give me a poke @alek-kam-robotec-ai when you'd like me to kick off a build +1
> > > 
> > > 
> > > @hultonha I've applied requested changes. Build should be ready to go. Thanks!
> > 
> > 
> > Awesome stuff, just kicked off a build @alek-kam-robotec-ai +1 https://jenkins.build.o3de.org/view/Projects/job/o3de-extras/view/change-requests/job/PR-101/1/
> 
> Although the build has failed it is not clear to me why. The errors inside the pipeline.log file do not seem to be connected in any way to the gem's code. Similar build results can be seen in the [Introduced XACRO parsing to ROS2 Gem](https://github.com/o3de/o3de-extras/commit/7ca4fb7f58b25786a71fba17c9ae731c006c8f47) commit (the worklfow run 2 hours before this one).

Hmm interesting, and thanks for following-up @moraaar. This might be something we need @byrcolin's help with. You can try updating your branch and I can start another build just in case to check, otherwise it might need to be investigated further. Thanks!

adamdbrw commented:
@hultonha, @byrcolin  this PR is only blocked by the CI status, but we do not know why the fail is happening and it does not seem to be related. Could we unblock this? 

The PR is holding us from releasing the RGL (Robotec GPU Lidar) Gem for O3DE.

hultonha commented:
> @hultonha, @byrcolin this PR is only blocked by the CI status, but we do not know why the fail is happening and it does not seem to be related. Could we unblock this?
> 
> The PR is holding us from releasing the RGL (Robotec GPU Lidar) Gem for O3DE.

I've started a new build just in case the error before was a one-off thing. Tagging @o3de/sig-build (@amzn-changml) if there's any more info about what might be causing this failure. Tagging @byrcolin as well for visibility.

In the previous log (https://jenkins.build.o3de.org/blue/organizations/jenkins/o3de-extras/detail/PR-101/1/pipeline/189) I see 

```
[2023-01-26T16:09:03.945Z] fatal: bad object d45b5ce140b830f1fe811cfc707cf6dc07986f7b
``` 

which doesn't look good... any idea what might be causing this?

michalpelka commented:
I can see in both builds one more clue:
https://jenkins.build.o3de.org/blue/organizations/jenkins/o3de-extras/detail/PR-101/1/pipeline/189/
https://jenkins.build.o3de.org/blue/organizations/jenkins/o3de-extras/detail/PR-101/2/pipeline/189/


```
[2023-01-31T11:18:58.781Z] D:\workspace>D:\workspace\o3de\scripts\o3de register --gem-path D:\workspace\o3de-extras\Gems\OpenXRVk   || ver1>nul 
[Pipeline] // node
[Pipeline] }
[2023-01-31T11:18:58.886Z] ERROR: Execution failed
[2023-01-31T11:18:58.897Z] java.lang.NullPointerException: Cannot invoke method containsKey() on null object
[2023-01-31T11:18:58.897Z] 	at org.codehaus.groovy.runtime.NullObject.invokeMethod(NullObject.java:91)
[2023-01-31T11:18:58.897Z] 	at org.codehaus.groovy.runtime.callsite.PogoMetaClassSite.call(PogoMetaClassSite.java:47)
[2023-01-31T11:18:58.897Z] 	at org.codehaus.groovy.runtime.callsite.CallSiteArray.defaultCall(CallSiteArray.java:47)
[2023-01-31T11:18:58.897Z] 	at org.codehaus.groovy.runtime.callsite.NullCallSite.call(NullCallSite.java:34)
[2023-01-31T11:18:58.897Z] 	at org.codehaus.groovy.runtime.callsite.CallSiteArray.defaultCall(CallSiteArray.java:47)
[2023-01-31T11:18:58.897Z] 	at org.codehaus.groovy.runtime.callsite.AbstractCallSite.call(AbstractCallSite.java:116)
[2023-01-31T11:18:58.897Z] 	at org.kohsuke.groovy.sandbox.impl.Checker$1.call(Checker.java:163)
[2023-01-31T11:18:58.897Z] 	at org.kohsuke.groovy.sandbox.impl.Checker.checkedCall(Checker.java:165)
[2023-01-31T11:18:58.897Z] 	at com.cloudbees.groovy.cps.sandbox.SandboxInvoker.methodCall(SandboxInvoker.java:17)
[2023-01-31T11:18:58.897Z] 	at WorkflowScript.CreateSingleNode(WorkflowScript:1008)
[2023-01-31T11:18:58.897Z] 	at ___cps.transform___(Native Method)
[2023-01-31T11:18:58.897Z] 	at com.cloudbees.groovy.cps.impl.ContinuationGroup.methodCall(ContinuationGroup.java:86)
[2023-01-31T11:18:58.897Z] 	at com.cloudbees.groovy.cps.impl.FunctionCallBlock$ContinuationImpl.dispatchOrArg(FunctionCallBlock.java:113)
[2023-01-31T11:18:58.897Z] 	at com.cloudbees.groovy.cps.impl.FunctionCallBlock$ContinuationImpl.fixArg(FunctionCallBlock.java:83)
[2023-01-31T11:18:58.897Z] 	at jdk.internal.reflect.GeneratedMethodAccessor477.invoke(Unknown Source)
[2023-01-31T11:18:58.897Z] 	at java.base/jdk.internal.reflect.DelegatingMethodAccessorImpl.invoke(DelegatingMethodAccessorImpl.java:43)
[2023-01-31T11:18:58.897Z] 	at java.base/java.lang.reflect.Method.invoke(Method.java:566)
[2023-01-31T11:18:58.897Z] 	at com.cloudbees.groovy.cps.impl.ContinuationPtr$ContinuationImpl.receive(ContinuationPtr.java:72)
[2023-01-31T11:18:58.897Z] 	at com.cloudbees.groovy.cps.impl.ConstantBlock.eval(ConstantBlock.java:21)
[2023-01-31T11:18:58.897Z] 	at com.cloudbees.groovy.cps.Next.step(Next.java:83)
[2023-01-31T11:18:58.897Z] 	at com.cloudbees.groovy.cps.Continuable$1.call(Continuable.java:174)
[2023-01-31T11:18:58.897Z] 	at com.cloudbees.groovy.cps.Continuable$1.call(Continuable.java:163)
[2023-01-31T11:18:58.897Z] 	at org.codehaus.groovy.runtime.GroovyCategorySupport$ThreadCategoryInfo.use(GroovyCategorySupport.java:136)
[2023-01-31T11:18:58.897Z] 	at org.codehaus.groovy.runtime.GroovyCategorySupport.use(GroovyCategorySupport.java:275)
[2023-01-31T11:18:58.897Z] 	at com.cloudbees.groovy.cps.Continuable.run0(Continuable.java:163)
[2023-01-31T11:18:58.897Z] 	at org.jenkinsci.plugins.workflow.cps.SandboxContinuable.access$001(SandboxContinuable.java:18)
[2023-01-31T11:18:58.897Z] 	at org.jenkinsci.plugins.workflow.cps.SandboxContinuable.run0(SandboxContinuable.java:51)
[2023-01-31T11:18:58.897Z] 	at org.jenkinsci.plugins.workflow.cps.CpsThread.runNextChunk(CpsThread.java:185)
[2023-01-31T11:18:58.897Z] 	at org.jenkinsci.plugins.workflow.cps.CpsThreadGroup.run(CpsThreadGroup.java:403)
[2023-01-31T11:18:58.897Z] 	at org.jenkinsci.plugins.workflow.cps.CpsThreadGroup.access$400(CpsThreadGroup.java:97)
[2023-01-31T11:18:58.897Z] 	at org.jenkinsci.plugins.workflow.cps.CpsThreadGroup$2.call(CpsThreadGroup.java:315)
[2023-01-31T11:18:58.897Z] 	at org.jenkinsci.plugins.workflow.cps.CpsThreadGroup$2.call(CpsThreadGroup.java:279)
[2023-01-31T11:18:58.897Z] 	at org.jenkinsci.plugins.workflow.cps.CpsVmExecutorService$2.call(CpsVmExecutorService.java:67)
[2023-01-31T11:18:58.897Z] 	at java.base/java.util.concurrent.FutureTask.run(FutureTask.java:264)
[2023-01-31T11:18:58.897Z] 	at hudson.remoting.SingleLaneExecutorService$1.run(SingleLaneExecutorService.java:139)
[2023-01-31T11:18:58.897Z] 	at jenkins.util.ContextResettingExecutorService$1.run(ContextResettingExecutorService.java:28)
[2023-01-31T11:18:58.897Z] 	at jenkins.security.ImpersonatingExecutorService$1.run(ImpersonatingExecutorService.java:68)
[2023-01-31T11:18:58.897Z] 	at java.base/java.util.concurrent.Executors$RunnableAdapter.call(Executors.java:515)
[2023-01-31T11:18:58.897Z] 	at java.base/java.util.concurrent.FutureTask.run(FutureTask.java:264)
[2023-01-31T11:18:58.897Z] 	at java.base/java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1128)
[2023-01-31T11:18:58.897Z] 	at java.base/java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:628)
[2023-01-31T11:18:58.897Z] 	at java.base/java.lang.Thread.run(Thread.java:829)
[2023-01-31T11:18:58.897Z] Retrying
[2023-01-31T11:18:58.917Z] 
```
It looks that Jenkins expects something that is not given by Gem registration.

hultonha commented on Gems/ROS2/Code/Include/ROS2/Lidar/LidarRaycasterBus.h:
```diff
@@ -0,0 +1,190 @@
+    //! Class used for creating typesafe Uuid types.
+    //! It utilizes the phantom types technique.
+    template<typename Tag>
+    class StronglyTypedUuid
```
This is super cool, thanks for adding this! 🙂

hultonha commented:
Thanks for the follow-up @michalpelka, I'll ping the relevant team about this now.

lumberyard-employee-dm commented on Gems/ROS2/Code/Include/ROS2/Lidar/LidarRaycasterBus.h:
```diff
@@ -0,0 +1,190 @@
+    {
+    public:
+        StronglyTypedUuid() = default;
+        StronglyTypedUuid(AZ::Uuid value)
+            : m_uuid(value)
+        {
+        }
```
This constructor should be marked explicit.
In order for this class to be strongly-typed it shouldn't allow implicit conversion from a `Uuid` to a `StronglyTypedUuid<Tag>` since it will allow for untended conversions to StronglyTypedUuids from plain Uuids.
```suggestion
        explicit StronglyTypedUuid(AZ::Uuid value)
            : m_uuid(value)
        {
        }
```

lumberyard-employee-dm commented on Gems/ROS2/Code/Include/ROS2/Lidar/LidarRaycasterBus.h:
```diff
@@ -0,0 +1,190 @@
+
+        bool IsNull() const
+        {
+            return m_uuid.IsNull();
+        }
+
+        bool operator==(const StronglyTypedUuid& rhs) const
+        {
+            return m_uuid == rhs.m_uuid;
+        }
+
+        bool operator!=(const StronglyTypedUuid& rhs) const
+        {
+            return m_uuid != rhs.m_uuid;
+        }
+
+        bool operator<(const StronglyTypedUuid& rhs) const
+        {
+            return m_uuid < rhs.m_uuid;
+        }
+
+        bool operator>(const StronglyTypedUuid& rhs) const
+        {
+            return m_uuid > rhs.m_uuid;
+        }
+
+        bool operator<=(const StronglyTypedUuid& rhs) const
+        {
+            return m_uuid <= rhs.m_uuid;
+        }
+
+        bool operator>=(const StronglyTypedUuid& rhs) const
+        {
+            return m_uuid >= rhs.m_uuid;
+        }
+
+        size_t GetHash() const
+        {
+            return m_uuid.GetHash();
+        }
```
All these functions can be marked `constexpr`

hultonha commented:
@michalpelka I believe this fix from @amzn-changml should hopefully resolve the build issue we were seeing before, I've just started another build now

hultonha commented:
@alek-kam-robotec-ai Are you okay to make the changes suggested by @lumberyard-employee-dm and then I'll start another build shortly, thanks!

hultonha commented:
(we could also do a separate PR as a fast follow-up too if that would be easier that just include the fixes to `StronglyTypedUuid`)

adamdbrw commented:
We need to submit PRs to demo projects. Hopefully we can do Gem versioning after the first release but this affects Lidar interfaces.

michalpelka reviewed (changes requested):
I've tried this PR against the ROSCon demo and it breaks the deserialization of prefabs there.
We need to check it and provide PRs to RosconDemo, the project template, and the Vacuum sample.

hultonha commented:
Good catch @michalpelka, let me know when those updates are made and I can start builds for those repos as needed.

Perhaps in the call later we can discuss the need for pinning a version to make development a little more stable for while?

michalpelka commented:
Found an issue with an uninitialized string, and rebased it. With ROSConDemo and RosRobotSample works fine, without changes with prefabs.

adamdbrw commented on Gems/ROS2/Code/Source/Lidar/ROS2LidarSensorComponent.cpp:
```diff
@@ -89,9 +89,20 @@ namespace ROS2
     }
+    AZStd::string GetDeafultLidarSystem(){
+        const auto& lidarSystemList = LidarRegistrarInterface::Get()->GetRegisteredLidarSystems();
+        AZ_Assert(!lidarSystemList.empty(), "Unable to choose a lidarSystem for the lidar sensor.");
```
Consider:
"No LIDAR system for the sensor to use" 
Also, assert will be gone in release build so this .front() call will still have a probelm

adamdbrw commented on Gems/ROS2/Code/Source/Lidar/ROS2LidarSensorComponent.cpp:
```diff
@@ -89,9 +89,20 @@ namespace ROS2
+    AZStd::string GetDeafultLidarSystem(){
+        const auto& lidarSystemList = LidarRegistrarInterface::Get()->GetRegisteredLidarSystems();
+        AZ_Assert(!lidarSystemList.empty(), "Unable to choose a lidarSystem for the lidar sensor.");
+        return lidarSystemList.front();
```
Could still be called on an empty list for undefined behavior in release build.

michalpelka replied:
Replace assert with a warning. Add an if clause to prevent UB.

adamdbrw commented on Gems/ROS2/Code/Source/Lidar/ROS2LidarSensorComponent.cpp:
```diff
@@ -89,9 +89,20 @@ namespace ROS2
             }
         }
     }
+    AZStd::string GetDeafultLidarSystem(){
```
GetDefaultLidarSystem

adamdbrw commented on Gems/ROS2/Code/Source/Lidar/ROS2LidarSensorComponent.cpp:
```diff
@@ -89,9 +89,20 @@ namespace ROS2
+        {
+            m_lidarSystem = GetDeafultLidarSystem();
+        }
+        const auto *lidarMetaData= LidarRegistrarInterface::Get()->GetLidarSystemMetaData(m_lidarSystem);
```
formatting - use reformat for this commit

adamdbrw commented on Gems/ROS2/Code/Source/RobotImporter/xacro/XacroUtils.cpp:
```diff
@@ -7,10 +7,13 @@
 #include <AzFramework/Process/ProcessCommunicator.h>
 #include <AzFramework/Process/ProcessWatcher.h>
+#include <QString>
 namespace ROS2::Utils::xacro
```
nit: spacing (keep empty line between includes and namespace)

adamdbrw commented on Gems/ROS2/Code/Source/Lidar/ROS2LidarSensorComponent.cpp:
```diff
@@ -89,9 +89,20 @@ namespace ROS2
+            m_lidarSystem = GetDeafultLidarSystem();
+        }
+        const auto *lidarMetaData= LidarRegistrarInterface::Get()->GetLidarSystemMetaData(m_lidarSystem);
+        AZ_Assert(lidarMetaData, "No metadata for \"%s\"", m_lidarSystem.c_str() );
```
again, how do we handle this in a release build?

adamdbrw commented on Gems/ROS2/Code/Source/RobotImporter/xacro/XacroUtils.cpp:
```diff
@@ -7,10 +7,13 @@
 #include <AzCore/XML/rapidxml.h>
 #include <AzFramework/Process/ProcessCommunicator.h>
 #include <AzFramework/Process/ProcessWatcher.h>
+#include <QString>
```
QString included twice. Also, this seems not relevant to the lidar PR

michalpelka replied:
Opened issue for changes in this file https://github.com/o3de/o3de-extras/issues/122

hultonha commented:
Let me know when the comments have been addressed @michalpelka and @alek-kam-robotec-ai and I can start a build 👍 Thanks!

adamdbrw commented:
@hultonha I believe this is a Go now, thank you all for going through the review process!

hultonha commented:
> @hultonha I believe this is a Go now, thank you all for going through the review process!

Excellent @adamdbrw, build started - https://jenkins.build.o3de.org/view/Projects/job/o3de-extras/view/change-requests/job/PR-101/4/

michalpelka commented on Gems/ROS2/Code/Source/Lidar/LidarSystem.h:
```diff
@@ -0,0 +1,34 @@
+        void Deactivate();
+
+    private:
+        static constexpr const char* SystemName = "PhysX";
```
Changed to Scene Queries.

danielemorra98 commented on Gems/ROS2/Code/Source/Lidar/LidarRaycaster.cpp:
```diff
@@ -5,49 +5,98 @@
 #include <AzFramework/Physics/Shape.h>
+#include <Lidar/LidarRaycaster.h>
+#include <Lidar/LidarTemplateUtils.h>
 
```
```suggestion 
#include <AzCore/std/smart_ptr/make_shared.h>

```

danielemorra98 commented on Gems/ROS2/Code/Source/Lidar/LidarRegistrarSystemComponent.cpp:
```diff
@@ -0,0 +1,110 @@
+ */
+
+#include <Lidar/LidarRegistrarSystemComponent.h>
+
```
```suggestion
#include <AzCore/Serialization/EditContext.h>
#include <AzCore/Serialization/SerializeContext.h>

```

danielemorra98 commented on Gems/ROS2/Code/Source/Lidar/LidarTemplate.cpp:
```diff
@@ -6,14 +6,49 @@
-#include <AzCore/Serialization/EditContext.h>
-#include <AzCore/Serialization/EditContextConstants.inl>
+#include <Lidar/LidarTemplate.h>
 
```
```suggestion
#include <AzCore/Serialization/EditContext.h>
#include <AzCore/Serialization/SerializeContext.h>

```

danielemorra98 commented on Gems/ROS2/Code/Source/Lidar/LidarTemplateUtils.cpp:
```diff
@@ -6,33 +6,134 @@
-#include <AzCore/Utils/Utils.h>
-#include <AzCore/std/containers/unordered_map.h>
+#include <Lidar/LidarTemplateUtils.h>
 
```
```suggestion
#include <AzCore/Math/Quaternion.h>

```

danielemorra98 reviewed (commented):
I built this against the [physics-test Project](https://github.com/o3de/o3de-physics-test-scene) and I got some compilation errors. It looked like a lack of dependencies

alek-kam-robotec-ai commented:
> I built this against the [physics-test Project](https://github.com/o3de/o3de-physics-test-scene) and I got some compilation errors. It looked like a lack of dependencies

I've addressed these comments in https://github.com/o3de/o3de-extras/pull/168.
